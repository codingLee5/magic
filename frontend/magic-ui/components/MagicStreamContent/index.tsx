import { useTyping } from "./useTyping"
import { useUpdateEffect } from "ahooks"
import type { HTMLAttributes, ReactNode } from "react"
import { memo } from "react"

export interface MagicStreamContentProps extends Omit<HTMLAttributes<HTMLDivElement>, "children"> {
	content: string
	children?: (content: string) => ReactNode
}

const MagicStreamContent = memo(function MagicStreamContent({
	content,
	children,
	...props
}: MagicStreamContentProps) {
	const { content: typingContent, add } = useTyping(content)

	useUpdateEffect(() => {
		add(content)
	}, [content, add])

	return <div {...props}>{children ? children(typingContent) : typingContent}</div>
})

export default MagicStreamContent
