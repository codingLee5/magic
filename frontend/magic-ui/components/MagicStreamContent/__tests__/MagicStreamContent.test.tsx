import { render, screen } from "@testing-library/react"
import MagicThemeProvider from "../../ThemeProvider"
import { describe, it, expect } from "vitest"
import MagicStreamContent from "../index"

const renderWithTheme = (component: React.ReactElement) =>
	render(<MagicThemeProvider theme="light">{component}</MagicThemeProvider>)

describe("MagicStreamContent", () => {
	it("应该正常渲染", () => {
		renderWithTheme(<MagicStreamContent content="流式内容" />)
		expect(screen.getByText("流式内容")).toBeInTheDocument()
	})

	// 快照测试
	describe("快照测试", () => {
		it("基础流式内容快照", () => {
			const { asFragment } = renderWithTheme(<MagicStreamContent content="流式内容" />)
			expect(asFragment()).toMatchSnapshot()
		})

		it("长文本流式内容快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicStreamContent content="这是一个很长的流式内容，用于测试组件的显示效果" />,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带自定义样式快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicStreamContent
					content="流式内容"
					style={{ fontSize: "16px", color: "blue" }}
				/>,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带类名快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicStreamContent content="流式内容" className="custom-stream" />,
			)
			expect(asFragment()).toMatchSnapshot()
		})

		it("空内容快照", () => {
			const { asFragment } = renderWithTheme(<MagicStreamContent content="" />)
			expect(asFragment()).toMatchSnapshot()
		})

		it("带HTML标签内容快照", () => {
			const { asFragment } = renderWithTheme(
				<MagicStreamContent content="<strong>粗体文本</strong>" />,
			)
			expect(asFragment()).toMatchSnapshot()
		})
	})
})
