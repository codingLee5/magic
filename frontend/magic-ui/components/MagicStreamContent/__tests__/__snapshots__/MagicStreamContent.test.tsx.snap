// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`MagicStreamContent > 快照测试 > 基础流式内容快照 1`] = `
<DocumentFragment>
  <div>
    流式内容
  </div>
</DocumentFragment>
`;

exports[`MagicStreamContent > 快照测试 > 带HTML标签内容快照 1`] = `
<DocumentFragment>
  <div>
    &lt;strong&gt;粗体文本&lt;/strong&gt;
  </div>
</DocumentFragment>
`;

exports[`MagicStreamContent > 快照测试 > 带类名快照 1`] = `
<DocumentFragment>
  <div
    class="custom-stream"
  >
    流式内容
  </div>
</DocumentFragment>
`;

exports[`MagicStreamContent > 快照测试 > 带自定义样式快照 1`] = `
<DocumentFragment>
  <div
    style="font-size: 16px; color: blue;"
  >
    流式内容
  </div>
</DocumentFragment>
`;

exports[`MagicStreamContent > 快照测试 > 空内容快照 1`] = `
<DocumentFragment>
  <div />
</DocumentFragment>
`;

exports[`MagicStreamContent > 快照测试 > 长文本流式内容快照 1`] = `
<DocumentFragment>
  <div>
    这是一个很长的流式内容，用于测试组件的显示效果
  </div>
</DocumentFragment>
`;
